cmd/skaffold/app/cmd/diagnose.go

```diff
@@ -0,0 +1,130 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"time"
+
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/docker"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/runner"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/schema/latest"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/version"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/watch"
+
+	"github.com/pkg/errors"
+	"github.com/spf13/cobra"
+)
+
+// NewCmdDiagnose describes the CLI command to diagnose skaffold.
+func NewCmdDiagnose(out io.Writer) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "diagnose",
+		Short: "Run a diagnostic on Skaffold",
+		Args:  cobra.NoArgs,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return doDiagnose(out)
+		},
+	}
+	cmd.Flags().StringVarP(&opts.ConfigurationFile, "filename", "f", "skaffold.yaml", "Filename or URL to the pipeline file")
+	return cmd
+}
+
+func doDiagnose(out io.Writer) error {
+	_, config, err := newRunner(opts)
+	if err != nil {
+		return errors.Wrap(err, "creating runner")
+	}
+
+	fmt.Fprintln(out, "Skaffold version:", version.Get().GitCommit)
+	fmt.Fprintln(out, "Configuration version:", config.APIVersion)
+	fmt.Fprintln(out, "Number of artifacts:", len(config.Build.Artifacts))
+
+	fmt.Fprintln(out, "\nRunning a diagnostic on artifacts and their dependencies:")
+
+	if err := diagnoseArtifacts(out, config.Build.Artifacts); err != nil {
+		return errors.Wrap(err, "running diagnostic on artifacts")
+	}
+
+	fmt.Fprintln(out, "\nSecond run (that benefits from cache):")
+
+	if err := diagnoseArtifacts(out, config.Build.Artifacts); err != nil {
+		return errors.Wrap(err, "running second diagnostic on artifacts")
+	}
+
+	return nil
+}
+
+func diagnoseArtifacts(out io.Writer, artifacts []*latest.Artifact) error {
+	for _, artifact := range artifacts {
+		start := time.Now()
+		deps, err := runner.DependenciesForArtifact(artifact)
+		if err != nil {
+			return errors.Wrap(err, "listing artifact dependencies")
+		}
+
+		fmt.Fprintln(out, "Type:", typeOfArtifact(artifact))
+		fmt.Fprintln(out, " - Dependencies to watch:", len(deps))
+		fmt.Fprintln(out, " - Time to list dependencies:", time.Since(start))
+
+		if artifact.DockerArtifact != nil {
+			size, err := sizeOfTheDockerContext(artifact)
+			if err != nil {
+				return errors.Wrap(err, "computing the size of the Docker context")
+			}
+
+			fmt.Fprintln(out, " - Size of the Docker context:", size)
```

Comment:
do we want a human readable size format here, or there's a reason you keep bytes? Even if so, I think a `bytes` unit can be useful.

Reply:
there are packages out there that will do automatic "humanization" of byte sizes for you, e.g. https://github.com/dustin/go-humanize